File: coding-puzzles/googleChallenges/search_words.js
function answer(document, searchTerms){
  var output = [];
  var docWordArray = document.split(' ');
  // Loop through each word in the doc
  for(i=0; i<docWordArray.length; i++){
    var snippetArray = [];


    if(searchTerms.indexOf(docWordArray[i]) >= 0){
      var matchArray = [];
      searchTerms.forEach(function(word){
        matchArray.push(word)
      });

      for(j=i; j<docWordArray.length; j++){
        if(matchArray.length > 0){
          if(searchTerms.indexOf(docWordArray[j]) >= 0){
            matchArray.splice( matchArray.indexOf(docWordArray[j]), 1 );
          }
          snippetArray.push(docWordArray[j]);
        }
      }
      if(matchArray.length !== 0) snippetArray = [];
    }


    // If there is a match or if the snippet array is shorter than the
    // current word array
    if(output.length == 0 || ( output.length > snippetArray.length && snippetArray.length > 0)){
      output = snippetArray;
    }
  }


  return output; //join it
}

console.log(answer("hi my name is nick duh nick my break my name nick this up is hi", ["nick", "is", "my"]))

File: coding-puzzles/googleChallenges/search_words.py
def answer(document, searchTerms):
    # Split up the doc into a list, declare the shortest snippet var
    shortestString = []
    docWordArray = document.split()

    # Loop through each word in the document and if the current word is
    # one of the search terms, set variables for a snippet array and
    # a word array to compare with the inputted search term array
    for i in range( 0, len(docWordArray) ):
        if docWordArray[i] in searchTerms:
            snippetArray = []
            searchTermCheck = []

            # Go through the document starting at the given word
            for j in range( i, len(docWordArray) ):

                # As long as the search term array is longer than the
                # search term check array, push words to the snippet
                if set(searchTerms) != set(searchTermCheck):
                    snippetArray.append(docWordArray[j])

                    # If the words are one of the search terms, push
                    # them to the check array so that the function knows
                    # when to stop adding to the snippet
                    if docWordArray[j] in searchTerms:
                        searchTermCheck.append(docWordArray[j])

            # If there is no shortest string or if the new snippet is
            # shorter than the current snippet AND has all the search
            # terms, set the new snippet to the shortest
            if len(shortestString) == 0:
                shortestString = snippetArray
            elif len(shortestString) > len(snippetArray) and set(searchTerms) == set(searchTermCheck):
                shortestString = snippetArray

    # Make it a string a return
    return " ".join(shortestString)



print answer("hi my name is nick duh nick my break my name nick is the the the my this up is hi", ["nick", "is", "my"])
